riscv-qemu: actually sleep in time.Sleep()

aykevl opened the pull request:
Instead of just incrementing the timestamp, this causes the system to actually sleep when calling time.Sleep. The direct effect is that this works as expected:

    $ tinygo run -target=riscv-qemu examples/serial
    hello world!
    hello world!
    hello world!
    [..etc]

This commit also adds a bare bones handler for exceptions (such as invalid memory writes), since we're adding an interrupt handler anyway.

While this patch doesn't add that much functionality, having interrupt support is going to be needed for multicore support on riscv-qemu. My plan is to first add this support to riscv-qemu (based on the earlier work I did for the RP2040 and demoed at FOSDEM 2025) and once the basics are in place and fully tested we can extend this support to the RP2040. Writing for QEMU first makes it much easier to debug any issues that will come up.

aykevl committed:
riscv-qemu: actually sleep in time.Sleep()

Instead of just incrementing the timestamp, this causes the system to
actually sleep when calling time.Sleep. The direct effect is that this
works as expected:

    $ tinygo run -target=riscv-qemu examples/serial
    hello world!
    hello world!
    hello world!
    [..etc]

This commit also adds a bare bones handler for exceptions (such as
invalid memory writes), since we're adding an interrupt handler anyway.

While this patch doesn't add that much functionality, having interrupt
support is going to be needed for multicore support on riscv-qemu. My
plan is to first add this support to riscv-qemu (based on the earlier
work I did for the RP2040 and demoed at FOSDEM 2025) and once the basics
are in place and fully tested we can extend this support to the RP2040.
Writing for QEMU first makes it much easier to debug any issues that
will come up.

eliasnaur commented on src/runtime/runtime_tinygoriscv_qemu.go:
```diff
@@ -13,30 +13,94 @@ import (
+
 func ticksToNanoseconds(ticks timeUnit) int64 {
-	return int64(ticks)
+	return int64(ticks) * 100
```
Nit: explain the magic number or name the constant?

eliasnaur commented on src/runtime/runtime_tinygoriscv_qemu.go:
```diff
@@ -13,30 +13,94 @@ import (
 
 func nanosecondsToTicks(ns int64) timeUnit {
-	return timeUnit(ns)
+	return timeUnit(ns / 100)
```
Ditto.

eliasnaur commented on src/runtime/runtime_tinygoriscv_qemu.go:
```diff
@@ -13,30 +13,94 @@ import (
+			timerWakeup.Set(1)
+			// Disable the timer, to avoid triggering the interrupt right after
+			// this interrupt returns.
+			riscv.MIE.ClearBits(1 << 7) // MTIE bit
```
Nit: name 1<<7?

eliasnaur commented on src/runtime/runtime_tinygoriscv_qemu.go:
```diff
@@ -13,30 +13,94 @@ import (
+	if cause&(1<<31) != 0 {
+		// Topmost bit is set, which means that it is an interrupt.
+		switch code {
+		case 7: // Machine timer interrupt
```
Nit: name the constant?

eliasnaur commented on src/runtime/runtime_tinygoriscv_qemu.go:
```diff
@@ -13,30 +13,94 @@ import (
+	// Enable the timer.
+	target := uint64(ticks() + d)
+	aclintMTIMECMP.Set(target)
+	riscv.MIE.SetBits(1 << 7) // MTIE
```
Ditto.

aykevl committed:
riscv: define CSR constants and use them where possible

aykevl commented:
@eliasnaur I've added the constants (or explained what they mean in the case of ticks). I've also added the same constants in the other RISC-V chips: fe310, k210, and esp32c3.

eliasnaur commented:
LGTM, thanks!

deadprogram commented:
Thank you for working on this @aykevl it is also much more readable now. Thanks @eliasnaur for the reviews! Now merging.
